File: operator.java
// Time Complexity : O(n4^n)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {

    public List<List<Integer>> combinationSum(int[] candidates, int target) {

        List<List<Integer>> resultantList = new ArrayList<>();
        //base case
        if(candidates == null){
            return resultantList;
        }
        backtrack(resultantList, new ArrayList<>(), candidates, target, 0);	// recursively call the backtrack function
        return resultantList;
    }

    public void backtrack(List<List<Integer>> resultantList, List<Integer> tempResult, int[] candidates, int currSum, int start){

        if(currSum == 0){
            //add tempList to result list 
            resultantList.add(new ArrayList(tempResult));
            return;
        }

        for(int i=start; i<candidates.length; i++){
            
            if(candidates[i] <= currSum){
                tempResult.add(candidates[i]);
                backtrack(resultantList, tempResult, candidates, currSum - candidates[i], i); // with start index fixed, continue to find all the possible combinations 
                tempResult.remove(tempResult.size()-1);	// once the result is found, remove the recently added element to get the proper list
            }
        }
    }

}

File: Operator.py
class Solution:
    def helper(self, num, target, index, path, calculated, tail):
        # base
        if index == len(num):
            if calculated == target:
                self.result.append(''.join(path))
            return
        # logic
        for i in range(index, len(num)):
            if num[index] != '0' or index == i:
                curr = num[index:i+1]
                if index == 0:
                    path.append(curr)
                    self.helper(num, target, i + 1, path, calculated + int(curr), int(curr))
                    path.pop()
                else:
                    path.append('+')
                    path.append(num[i])
                    self.helper(num, target, i + 1, path, calculated + int(curr), int(curr))
                    path.pop()
                    path.pop()

                    path.append('-')
                    path.append(num[i])
                    self.helper(num, target, i + 1, path, calculated - int(curr), -1 * int(curr))
                    path.pop()
                    path.pop()

                    path.append('*')
                    path.append(num[i])
                    self.helper(num, target, i + 1, path, (calculated - tail) + (tail * int(curr)), tail * int(curr))
                    path.pop()
                    path.pop()

    def addOperators(self, num: str, target: int) -> list[str]:
        self.result = []
        self.helper(num, target, 0, [], 0, 0)
        return self.result


print(Solution().addOperators('123', 6))
